// vue-router/index.js
import {
  computed,
  defineComponent,
  h,
  inject,
  provide,
  reactive,
  readonly,
} from 'vue'

const routerKey = Symbol('router')
const routeKey = Symbol('route')
const depthKey = Symbol('router-depth')

const normalizePath = (path) => {
  if (!path) return '/'
  const withSlash = path.startsWith('/') ? path : `/${path}`
  return withSlash.replace(/\/+/g, '/').replace(/\/$/, '') || '/'
}

const joinPaths = (parent, child) => {
  if (!child) return normalizePath(parent)
  if (child.startsWith('/')) return normalizePath(child)
  return normalizePath(`${normalizePath(parent)}/${child}`)
}

const compileMatcher = (path) => {
  const keys = []
  const pattern = normalizePath(path)
    .replace(/:[^/]+/g, (segment) => {
      keys.push(segment.slice(1))
      return '([^/]+)'
    })
  return {
    keys,
    regex: new RegExp(`^${pattern}$`)
  }
}

const enrichRoute = (record, parent) => {
  const fullPath = joinPaths(parent?.fullPath || '/', record.path || '')
  const matcher = compileMatcher(fullPath)
  const enriched = {
    ...record,
    fullPath,
    matcher,
    meta: { ...(parent?.meta || {}), ...(record.meta || {}) }
  }

  enriched.children = (record.children || []).map((child) => enrichRoute(child, enriched))
  return enriched
}

const findMatch = (records, path, matched = [], params = {}) => {
  for (const record of records) {
    const result = record.matcher.regex.exec(path)
    const nextMatched = [...matched, record]

    if (result) {
      const nextParams = { ...params }
      record.matcher.keys.forEach((key, index) => {
        nextParams[key] = decodeURIComponent(result[index + 1])
      })

      const childMatch = findMatch(record.children || [], path, nextMatched, nextParams)
      return childMatch || { matched: nextMatched, params: nextParams }
    }

    const childOnlyMatch = findMatch(record.children || [], path, nextMatched, params)
    if (childOnlyMatch) return childOnlyMatch
  }

  return null
}

const parseLocation = (fullPath) => {
  const url = new URL(fullPath, window.location.origin)
  return {
    path: normalizePath(url.pathname),
    fullPath: `${normalizePath(url.pathname)}${url.search}`,
    query: Object.fromEntries(url.searchParams.entries())
  }
}

export const createWebHistory = () => ({ type: 'web' })

export const createRouter = ({ routes }) => {
  const records = routes.map((route) => enrichRoute(route))
  const currentRoute = reactive({ path: '/', fullPath: '/', query: {}, params: {}, matched: [], meta: {} })
  const beforeHooks = []

  const resolve = (target) => {
    if (typeof target === 'string') {
      const location = parseLocation(target)
      const match = findMatch(records, location.path)
      return {
        ...location,
        params: match?.params || {},
        matched: match?.matched || [],
        meta: match?.matched?.[match.matched.length - 1]?.meta || {}
      }
    }

    if (target?.name) {
      const match = records
        .flatMap((record) => [record, ...(record.children || [])])
        .find((record) => record.name === target.name)
      if (!match) return resolve('/')
      let path = match.fullPath
      Object.entries(target.params || {}).forEach(([key, value]) => {
        path = path.replace(`:${key}`, value)
      })
      const queryString = new URLSearchParams(target.query || {}).toString()
      return resolve(queryString ? `${path}?${queryString}` : path)
    }

    return resolve('/')
  }

  const runBeforeEach = (to, from, done) => {
    let index = 0
    const next = (override) => {
      if (typeof override === 'undefined') {
        const hook = beforeHooks[index++]
        if (hook) {
          hook(to, from, next)
        } else {
          done(to)
        }
        return
      }

      if (override && typeof override === 'object') {
        done(resolve(override))
        return
      }

      done(to)
    }

    next()
  }

  const push = (target) => {
    const from = { ...currentRoute }
    const to = resolve(target)

    runBeforeEach(to, from, (finalRoute) => {
      window.history.pushState({}, '', finalRoute.fullPath)
      Object.assign(currentRoute, finalRoute)
    })
  }

  const replace = (target) => {
    const finalRoute = resolve(target)
    window.history.replaceState({}, '', finalRoute.fullPath)
    Object.assign(currentRoute, finalRoute)
  }

  const router = {
    currentRoute: readonly(currentRoute),
    options: { routes },
    resolve,
    push,
    replace,
    beforeEach: (hook) => beforeHooks.push(hook),
    install(app) {
      const updateCurrentRoute = () => {
        Object.assign(currentRoute, resolve(window.location.pathname + window.location.search))
      }

      updateCurrentRoute()
      window.addEventListener('popstate', updateCurrentRoute)

      app.provide(routerKey, router)
      app.provide(routeKey, readonly(currentRoute))
      app.component('RouterLink', RouterLink)
      app.component('RouterView', RouterView)
      app.config.globalProperties.$router = router
      app.config.globalProperties.$route = currentRoute
    }
  }

  return router
}

export const useRouter = () => inject(routerKey)
export const useRoute = () => inject(routeKey)

const RouterLink = defineComponent({
  name: 'RouterLink',
  props: {
    to: {
      type: [String, Object],
      required: true
    },
    class: {
      type: [String, Array, Object],
      default: ''
    }
  },
  setup(props, { slots, attrs }) {
    const router = useRouter()
    const route = inject(routeKey)
    const target = computed(() => router.resolve(props.to))
    const isActive = computed(() => route.path === target.value.path)

    return () =>
      h(
        'a',
        {
          ...attrs,
          href: target.value.fullPath,
          class: [props.class, { 'router-link-active': isActive.value }],
          onClick: (event) => {
            event.preventDefault()
            router.push(props.to)
          }
        },
        slots.default ? slots.default() : target.value.fullPath
      )
  }
})

const RouterView = defineComponent({
  name: 'RouterView',
  setup() {
    const route = inject(routeKey)
    const depth = inject(depthKey, 0)
    provide(depthKey, depth + 1)

    return () => {
      const record = route.matched[depth]
      if (!record?.component) return null
      return h(record.component)
    }
  }
})
